### DISCUSSION TITLE
xPacks automatic install of compiler and tools

### DISCUSSION THREAD

**Discussion by trampas** [eclipse-embed-cdt/eclipse-plugins#587] [Ideas]
I was wondering if it would be possible to have a 'packages.json' file in embedded C/C++ project which contains the xPacks tools and compiler to install.  Then when a user opens the project eclipse could install the tools and make sure paths are correct such that developers have minimal work setting up tools to build new projects. 

Thanks
Trampas

**Comment by jonahgraham** [comment]
I think that is the idea set out in  https://eclipse-embed-cdt.github.io/templates/hello-world-qemu-xpack/ - not sure if anything is missing to make it complete.

**Comment by ilg-ul** [comment]
Yes, that's exactly the point with xPacks, to provide a reproducible development environment.

The recent Embedded CDT plug-ins added the option to prefer the tools installed in `xpacks/.bin` to those installed in the traditional locations (there is a button in the project settings).

As Jonah mentioned, this configuration is already used by the projects created via the Hello World QEMU template.

However, you don't have to use the template, you can manually run `xpm init` to add a `package.json` to your project, and install the binary tools by invoking `xpm install @xpack-dev-tool/xxx` in the project folder.

Then you save the `package.json` to your repo; someone else can clone it, and invoke `xpm install` in the project folder to get exactly the same tools as you did originally. 

The nice thing is that this works on any platform (Windows, macOS, Linux).

**Comment by trampas** [comment]
It would be nice if/when eclipse can not find compiler or tools and the
project has a package.json it would either automatically run xpm install or
ask user if they want to run the install.

I am finding the technical overhead required for many developers to open
terminal and type xpm install is too difficult.  That is they expect they
can download project double click on project file and everything will
magically work.

I can understand this desire since many IDEs appear to do this it is.
xPacks is a great tool for the job!


Thanks
Trampas

**Comment by ilg-ul** [comment]
Please do not reply by email, but use the web interface. Or, at least, do not quote back the entire message. See [this](https://github.com/eclipse-embed-cdt/eclipse-plugins/discussions/587#discussioncomment-6817440).

> I am finding the technical overhead required for many developers to open
terminal and type xpm install is too difficult.

:-(

Given the need to install node/npm/xpm, your desire to avoid the terminal entirely is not realistic.

However, once these tools are installed, it is possible to add a menu entry to invoke `xpm install`. Same for `xpm init`.

**Reply by trampas** [reply]
I had assumed that since eclipse was using xPacks it was now installing node and xpm.   

Maybe their is a better solution to my problem, how do I make it where a person can download a git repo and ideally with one click open up and IDE and be able to cross compile a bit exact binary output from source file?    

To make this simple assume this is windows only, not to bash windows users, but if someone is using Mac or Linux, they are generally able to follow instructions and use a command prompt, and as such I only see this problem with windows users. 

Has someone else solved this problem? If so how? 
Is this possible with Eclipse? 
Is it possible for Eclipse to add capability?
Is there another IDE where this is possible? 

Note I do have CI with git, so it is not about generating the binary file,  it is about getting IDE working.    I also know getting the IDE working and dependencies installed is just the first hurdle, but it is still blocking for users.  

I am considering using VMs to solve this problem.  That is install a virtual machine with tools such that eclipse works.  Then just letting people use the VM.  This creates all kinds of issues, like what happens if you upgrade the compiler, how do you get everyone to update VMs, etc.   I know several companies which do this. They keep read only versions of VM on servers and then everyone gets a copy.   

I have also heard of people shipping a laptops with tools installed.  

I am considering if I could make a custom windows installer, which is new to me, that would install eclipse, node, npm, xpm, git, etc.   I am not sure how to do this so if anyone has suggestions I would be interested in knowing more. 

Again I am open to any suggestions or method others have used or thought of. 

Thanks
Trampas

**Reply by ilg-ul** [reply]
> I had assumed that since eclipse was using xPacks it was now installing node and xpm.

Nope. 

These are the steps: <https://xpack.github.io/install/>. If your users are not able to do this, I'm afraid that there is not much we can do for them.

> I have also heard of people shipping a laptops with tools installed.

Sure. Make those MacBooks! ;-) And when a toolchain update is needed, ship them new laptops! 

> I am considering if I could make a custom windows installer

The early toolchain packages included a Windows installer. In the end... lots of problems. Now the Windows toolchains are .zip archives, and xpm manages them pretty well.

**Reply by trampas** [reply]
> These are the steps: [xpack.github.io/install](https://xpack.github.io/install/). If your users are not able to do this, I'm afraid that there is not much we can do for them.

I completely agree that if they can not follow instructions this will not be the last hurdle they have.  However many can not follow these simple steps. 

> Sure. Make those MacBooks! ;-) And when an update is needed, ship them new laptops!

If they were smart enough to have and use a MacBook I would not have this problem.  However for some of my clients a MacBook is a valid option and within their budget.  That is having the problem solved is worth the cost of a MacBook.  

> The early toolchain packages included a Windows installer. In the end lots of problems. Now the Windows toolchains are .zip archives, and xpm manages them pretty well.

Yes I was thinking that if I packaged up eclipse and xpm  into one install. As part of normal build process I could run 'xpm install'.  That is every build it would run 'xpm install' such that if compiler or tools change they are updated.  

Note this problem with Eclipse is not unique to me. I have talked with many contractors and developers.  Most say they never use Eclipse because of this problem.  One engineer suffers using Segger Embedded Studio (SES) as it is a one click build of project .  He said that his team of 10+ all use SES because they could install IDE and then be building code with one click of project file.  He said they buy SES for each developer and it is worth it so he does not have to support setting up eclipse. 
 
I have talked to other developers who use eclipse, however they will also use another IDE like Segger Embedded Studio, IAR, MPLAB, etc.  That is they do their work in eclipse or VScode, but maintain another IDE project that also builds the code for novice developers.  This way they have the one click build and do not have to support for eclipse or VScode.  
I would rather have an IDE with no syntax highlighting or features but would build the project with bit exact binary than have eclipse.   The value of eclipse is not worth the support issues when you have novice developers.  On projects you have to support that lowest common denominator developer.  Worse still is often you have no control over hiring of developers, as such the bar is really low.  Many can not even handle the Arduino IDE.  

The ideal solution is to have an project file in you repo.  Such that in windows if you double click on the project file it would know to install eclipse along with xPacks, and then do an xpm install for tools and dependencies.  After which Eclipse opens the project and shows text with arrow saying "Click this button to build project".   

We can discuss how stupid the developers are that need this and how we don't want them in our code if they can not pass the IDE install test, etc.   I would rather discuss if this is possible and how to make it happen if it is possible. 

Thanks
Trampas

**Reply by ilg-ul** [reply]
> The ideal solution is to have an project file in you repo. ... Eclipse ... shows text with arrow saying "Click this button to build project".

Probably in the age of ChatGPT this is how things will work.

> I would rather discuss if this is possible and how to make it happen if it is possible.

Sure, we can discuss, but I have little to no experience in dealing with such unqualified developers.

The xPack project was my solution to automating multi-platform tests for µOS++ (you can read the [history page](https://xpack.github.io/project/history/; preferably follow the link to the µOS++ history too).

These multi-platform tests are pretty complex, (read _complicated_), and require a good technical knowledge.

It is true that beginners can also benefit from the automation provided by xpm, but they are not the main target audience.

So, if you think you can further package Eclipse, xpm and any other tools you need, sure, go for it, but I can't commit too many resources into this.

**Comment by TommyMurphyTM1234** [comment]
> I had assumed that since eclipse was using xPacks it was now installing node and xpm.

Just to be accurate, Eclipse does not use xPacks.
But you can use xPacks from Eclipse.

> I am considering if I could make a custom windows installer

FWIW I found Inno Setup easy to use for Windows installers.
And InstallBuilder where cross platform support was required.

* https://jrsoftware.org/isinfo.php
* https://installbuilder.com/

**Reply by ilg-ul** [reply]
> Just to be accurate, Eclipse does not use xPacks.

To be even more accurate, the Eclipse Embedded CDT uses xPacks, when available.

For example you can select the path of the toolchain by selecting the xPack version.

Also the build plugin has a setting to prefere the local tools installed in `xpack/.bin`.

Not to mention that the Hello World QEMU project template plug-in uses xpm to install dependencies.

So, in one form or another, Eclipse uses xPacks.

**Comment by jonahgraham** [comment]
The Eclipse Ecosystem already provides an embedded (shipped as Eclipse Plugins in a p2 site) of the node runtime that we can integrate. See https://github.com/eclipse-wildwebdeveloper/wildwebdeveloper/tree/master/org.eclipse.wildwebdeveloper.embedder.node for the guide.

That should allow the bootstrapping of xpm from within Eclipse entirely with no additional tool installation required. Similar to how other node dependent packages in the Eclipse ecosystem work, the xpm installtion can be done at "our" build time.

**Reply by ilg-ul** [reply]
> That should allow the bootstrapping of xpm from within Eclipse entirely with no additional tool installation required. Similar to how other node dependent packages in the Eclipse ecosystem work, the xpm installation can be done at "our" build time.

This might be an interesting path to investigate.

---

However this is a slightly different discussion, if Jonah thinks there is interest in it, we can open a separate topic.

I think that installing GCC as an xPack dependency can be useful, especially on platforms like Windows, for CDT too, not only for Embedded CDT.

For example I can add a build plug-in similar to the cross build plug-ins for arm/riscv, to manage native gcc/clang. 

The advantage would be that programs using this plug-in will work exactly the same on all platforms.

If I remember right, now CDT had multiple toolchain definitions, for Linux, macOS, Windows, and creating projects practically binds them to a platform.

**Reply by trampas** [reply]
I personally would love to have xPacks "native" with eclipse.

**Comment by trampas** [comment]
Erich Styger is appearing to have similar issues:
https://mcuoneclipse.com/2023/08/25/vs-code-virtual-environments-for-embedded-development-with-conda/
He is using VSCode and Conda for IDE and dependencies but still does not have the one click install.  

Conda looks like an interesting alternative to xPacks

**Reply by ilg-ul** [reply]
> Conda looks like an interesting alternative to xPacks

More or less. They address different requirements and are part of different ecosystems.

xpm is actually a **project** manager (which is more than a package manager), since it manages build configurations, actions, properties, substitutions, etc.

And it is planned to add a project builder, similar to Eclipse managed build, but based on components, like in eCos CDL.
